src/cdk-experimental/ui-patterns/behaviors/event-manager/event-manager.ts

```diff
@@ -0,0 +1,185 @@
+  protected afterFns: ((event: T) => void)[] = [];
+
+  protected defaultHandlerOptions: EventHandlerOptions = {
+    preventDefault: false,
+    stopPropagation: false,
+  };
+
+  constructor(defaultHandlerOptions?: Partial<EventHandlerOptions>) {
+    this.defaultHandlerOptions = {
+      ...this.defaultHandlerOptions,
+      ...defaultHandlerOptions,
+    };
+  }
+
+  /**
+   * Composes together multiple event managers into a single event manager that delegates to the
+   * individual managers.
+   */
+  static compose<T extends Event>(...managers: EventManager<T>[]) {
+    const composedManager = new GenericEventManager<T>();
+    composedManager._submanagers = managers;
+    return composedManager;
+  }
+
+  /**
+   * Runs any handlers that have been configured to handle this event. If multiple handlers are
+   * configured for this event, they are run in the order they were configured. Returns
+   * `true` if the event has been handled, otherwise returns `undefined`.
+   *
+   * Note: the use of `undefined` instead of `false` in the unhandled case is necessary to avoid
+   * accidentally preventing the default behavior on an unhandled event.
+   */
+  async handle(event: T): Promise<true | undefined> {
+    if (!this.isHandled(event)) {
+      return undefined;
+    }
+    for (const fn of this.beforeFns) {
+      fn(event);
+    }
+    for (const submanager of this._submanagers) {
+      await submanager.handle(event);
+    }
+    for (const config of this.getConfigs(event)) {
+      await config.handler(event);
+      if (config.stopPropagation) {
+        event.stopPropagation();
+      }
+      if (config.preventDefault) {
+        event.preventDefault();
+      }
+    }
+    for (const fn of this.afterFns) {
+      fn(event);
+    }
+    return true;
+  }
+
+  /**
+   * Configures the event manager to run a function immediately before it as about to handle
+   * any event.
+   */
+  beforeHandling(fn: (event: T) => void): this {
+    this.beforeFns.push(fn);
+    return this;
+  }
+
+  /**
+   * Configures the event manager to run a function immediately after it handles any event.
+   */
+  afterHandling(fn: (event: T) => void): this {
+    this.afterFns.push(fn);
+    return this;
+  }
+
+  /**
+   * Configures the event manager to handle specific events. (See subclasses for more).
+   */
+  abstract on(...args: [...unknown[]]): this;
+
+  /**
+   * Gets all of the handler configs that are applicable to the given event.
+   */
+  protected abstract getConfigs(event: T): EventHandlerConfig<T>[];
+
+  /**
+   * Checks whether this event manager is confugred to handle the given event.
+   */
+  protected isHandled(event: T): boolean {
+    return this.getConfigs(event).length > 0 || this._submanagers.some(sm => sm.isHandled(event));
+  }
+}
+
+/**
+ * A generic event manager that can work with any type of event.
+ */
+export class GenericEventManager<T extends Event> extends EventManager<T> {
```

Comment:
This class signature seems identical to just `EventManager`? Am I missing something?

Reply:
@mmalerba

Reply:
I forget why I did this, something about the typing with the generics I think. It can probably be cleaned up to not need this separate class